### ISSUE TITLE
README: Create a section on how to create a Workbench

### ISSUE THREAD

**Issue by luzpaz** [FreeCAD/FreeCAD-addons#104]
Relevant links: 

* https://github.com/FreeCAD/Workbench-Starterkit
* https://freecadweb.org/wiki/Workbench_creation

**Comment by luzpaz** [comment]
CC @looooo

**Comment by luzpaz** [comment]
It's not related ? 
Isn't it a template for a workbench ?
Addons are workbenches, modules, and macros... seems related to me.

**Comment by looooo** [comment]
yes it's a template for a work-bench / module. But I don't want to mix this with the old way of creating workbenches, as this will lead to misunderstandings. Currently both ways are supported, but I guess we should think about how to make both ways more compatible as the old-style workbenches won't go away. Also most users that trying to code something for freecad have not enough python background to understand why new way of creating workbenches is better.
But yes, I will try to work on documentation if time allows.

**Comment by pierrard** [comment]
Many thanks Lorenz  (Loooo) and other contributors for all you do for the community !!
I am slowly building a DIY router, 5 axis milling, and will very much appreciate the use of FreeCAD for the CAD and CAM once my hardware is ready (slow down by the cost of the parts now).
MANY THANKS!

**Comment by luzpaz** [comment]
@looooo I see, thanks for explaining that. I didn't understand there was a 'better' preference. IMHO, we could create documentation for this pre-0.18 so any of the users that 'come in to the fold' of FreeCAD can make a more educated decision.

**Comment by looooo** [comment]
@pierrard I am only doing what I need / like to work at. There are others like @luzpaz who try to move FreeCAD in the right direction. Best to thank the whole community, which include a lot of smart people fighting for a great project ; ).

@luzpaz yes you are right. I always underestimate the impact of docs : ).

**Comment by luzpaz** [comment]
@joelgraff perhaps you could weigh in on workbench creation as well at some point ? (i know your todo list is yuuuuge but nonetheless, please consider thanks!!!)

**Comment by joelgraff** [comment]
@luzpaz - Thanks for mentioning this.  Currently, I've [forked the starter kit](https://github.com/joelgraff/Workbench-Starterkit) and am, a little bit at a time, trying to build out  a general layout that is maintainable and scalable for small to mid-sized Python workbenches.  Specifically, I feel a workbench's functions are best divided across individual "Tasks" and code needs to be organized to reflect that. 

For example:

- `init_gui.py` should never have anything more than the command class imports and toolbar / context menu / menu layout code.
- A command class should only setup a task and clean up after it
- A base task class should do two things: initialize the task panel ui, and kick off the task-specific subclasses, where all of the custom code really belongs.

Of course, there are exceptions (like a command that simply executes an external program, or just opens a file), but for more complex tasks, I really think that sort of structure should be encouraged / enforced where possible.

Without a clear structure in place to mange the various workbench command / task use cases, we cannot write clear, useful developer documentation.  I realize it's no good trying to enforce absolute standards, but the goal is to provide a clean and well organized picture of how everything fits together rather than leave new devs to their own devices to hack together something that works, but may become unmaintainable as the workbench scope scales.

**Comment by luzpaz** [comment]
@joelgraff   
I resonate with everything you wrote and I'm very glad this discussion is progressing. Especially:  
> Without a clear structure in place to mange the various workbench command / task use cases, we cannot write clear, useful developer documentation. I realize it's no good trying to enforce absolute standards, but the goal is to provide a clean and well organized picture of how everything fits together rather than leave new devs to their own devices to hack together something that works, but may become unmaintainable as the workbench scope scales.

I think I'll pin this issue so it's of priority.

**Comment by luzpaz** [comment]
yikes, got to get around to doing this.

**Comment by joelgraff** [comment]
I had a framework of my own cobbled together, but I think I lost track of
it, too.
*Joel Graff, P.E.*

*LinkedIn:* https://www.linkedin.com/in/joelcgraff

**Comment by Eldo420** [comment]
Creating a detailed section in the README to explain how to create a Workbench for FreeCAD will help developers contribute effectively to the project. Here is an expanded guide that breaks down the process step by step:

### Step-by-Step Guide to Creating a Workbench in FreeCAD

1. **Introduction**

   This section explains the prerequisites, necessary tools, and a brief overview of what you will create.
   
   ```markdown
   ## Creating a Custom Workbench in FreeCAD

   FreeCAD provides a way to extend its functionalities through custom workbenches. This guide will walk you through the process of creating your own workbench.
   
   ### Prerequisites

   - Basic knowledge of Python
   - FreeCAD installed on your system
   - A text editor or IDE (e.g., Visual Studio Code, PyCharm)

   ```

2. **Setting Up Your Development Environment**

   Describe how to set up a development environment.

   ```markdown
   ### Setting Up Your Development Environment

   To get started with developing a custom Workbench, you need to ensure that your development environment is correctly set up. Follow the steps below:
   
   1. **Install FreeCAD**: Make sure FreeCAD is installed on your system.
   2. **Locate the `Mod` Directory**: Your custom Workbench will reside in the `Mod` directory of your FreeCAD installation. On Windows, it is usually located at `C:\Program Files\FreeCAD\Mod`. On Linux, it's typically found in `/usr/lib/freecad/Mod` or `/usr/local/lib/freecad/Mod`.
   3. **Create a New Folder for Your Workbench**: For example, create a folder named `MyWorkbench`.
   
   ```
   
3. **Basic Folder Structure**

   Outline the basic folder structure necessary for a FreeCAD Workbench.
   
   ```markdown
   ### Basic Folder Structure
   
   Your Workbench will follow a specific folder and file structure. Below is an example:
   
   ```
   MyWorkbench/
   ├── Init.py
   ├── InitGui.py
   ├── MyWorkbench.py
   ├── Resources/
       ├── icons/
       ├── translations/
   └── __init__.py
   ```
   
   - **`Init.py`**: Contains initialization code that runs when FreeCAD starts.
   - **`InitGui.py`**: Handles graphical user interface initialization.
   - **`MyWorkbench.py`**: The main file containing your workbench logic.
   - **`Resources`**: Contains icons, translation files, and other resources.
   - **`__init__.py`**: Indicates that the directory is a Python package.
   ```

4. **Creating Required Files**

   Detail the contents and purpose of each required file.

   ```markdown
   ### Creating Required Files

   #### Init.py

   This file will contain code to initialize your workbench when FreeCAD starts.
   
   ```python
   def initialize():
       """ This function is executed when FreeCAD starts """
       import MyWorkbench
       MyWorkbench.initialize()

   def FreeCADInit():
       """ This function is necessary for FreeCAD to recognize the Workbench """
       return "MyWorkbench"
   ```

   #### InitGui.py

   This file manages the GUI components of your Workbench.
   
   ```python
   class MyWorkbenchWorkbench (Workbench):
       """ The Workbench class provides necessary information """
       def __init__(self):
           self.__class__.MenuText = "My Workbench"
           self.__class__.ToolTip = "My custom workbench for FreeCAD"
           self.__class__.Icon = ":/icons/myworkbench_icon.svg"
           
       def Initialize(self):
           """ This function is executed when the Workbench gets activated """
           import MyWorkbenchGui
           self.appendToolbar("My Tools", ["MyTool"])
           self.appendMenu("My Workbench", ["MyTool"])

       def Activated(self):
           """ Execute this function when the workbench is loaded """
           from PySide import QtCore, QtGui

       def Deactivated(self):
           """ Execute this when switching away from the workbench """
           pass

       def GetClassName(self):
           return "Gui::PythonWorkbench"

   Gui.addWorkbench(MyWorkbenchWorkbench())
   ```

5. **Adding Tools**

   Explain how to create tools that will be part of the Workbench.

   ```markdown
   ### Adding Tools

   To add tools to your workbench, you'll need to create corresponding Python scripts and define them in the GUI initialization file.
   
   #### Example Tool Script: MyTool.py

   ```python
   class MyTool:
       """ Tool to perform a specific task """
       def Activated(self):
           print("MyTool Activated")
           
       def GetResources(self):
           return {
               'Pixmap' : 'path/to/your/tool/icon.svg',
               'MenuText': 'My Tool',
               'ToolTip' : 'Description of My Tool'
           }

   Gui.addCommand('MyTool', MyTool())
   ```

   #### Add the Tool to InitGui.py

   Update your `InitGui.py` to include the new tool in a toolbar or menu.
   
   ```python
   self.appendToolbar("My Tools", ["MyTool"])
   self.appendMenu("My Workbench", ["MyTool"])
   ```

6. **Resource Files**

   Explain where to place resource files and how to use them.

   ```markdown
   ### Resource Files

   Place your icons, translations, and other resources under the `Resources` directory. Use them in your tool and workbench definitions as shown in the scripts above.

   ```
   
7. **Packaging and Installing Your Workbench**

   Describe how to package the workbench for distribution and how to install it in FreeCAD.
   
   ```markdown
   ### Packaging and Installing Your Workbench

   - **Packaging**: Zip your `MyWorkbench` folder and its contents.
   - **Installing**: Unzip the folder into FreeCAD's `Mod` directory.

   When users start FreeCAD, your Workbench should now appear in the Workbench list.

   ```
   
8. **Advanced Features**

   Optionally include more advanced features like adding custom dialogs, using Qt for UI elements, or integrating complex algorithms.

   ```markdown
   ### Advanced Features

   - **Custom Dialogs**: Use `PySide` or `PyQt` to create custom dialogs.
   - **Complex Algorithms**: Integrate third-party Python libraries to add advanced capabilities.
   ```

9. **Troubleshooting and Debugging**

   Provide tips for troubleshooting common issues and debugging.

   ```markdown
   ### Troubleshooting and Debugging

   - **Log Files**: Check FreeCAD log files for any error messages.
   - **Debugging**: Use Python's `pdb` module or IDE debugging tools.
   ```

10. **Conclusion**

    Summarize the steps and encourage contributions.

    ```markdown
    ### Conclusion

    You've now created a basic custom Workbench for FreeCAD. Feel free to extend it with more features and share it with the community!
    ```

### Complete Example README Section

```markdown
## Creating a Custom Workbench in FreeCAD

FreeCAD provides a way to extend its functionalities through custom workbenches. This guide will walk you through the process of creating your own workbench.

### Prerequisites

- Basic knowledge of Python
- FreeCAD installed on your system
- A text editor or IDE (e.g., Visual Studio Code, PyCharm)

### Setting Up Your Development Environment

To get started with developing a custom Workbench, you need to ensure that your development environment is correctly set up. Follow the steps below:

1. **Install FreeCAD**: Make sure FreeCAD is installed on your system.
2. **Locate the `Mod` Directory**: Your custom Workbench will reside in the `Mod` directory of your FreeCAD installation. On Windows, it is usually located at `C:\Program Files\FreeCAD\Mod`. On Linux, it's typically found in `/usr/lib/freecad/Mod` or `/usr/local/lib/freecad/Mod`.
3. **Create a New Folder for Your Workbench**: For example, create a folder named `MyWorkbench`.

### Basic Folder Structure

Your Workbench will follow a specific folder and file structure. Below is an example:

```
MyWorkbench/
├── Init.py
├── InitGui.py
├── MyWorkbench.py
├── Resources/
    ├── icons/
    ├── translations/
└── __init__.py
```

- **`Init.py`**: Contains initialization code that runs when FreeCAD starts.
- **`InitGui.py`**: Handles graphical user interface initialization.
- **`MyWorkbench.py`**: The main file containing your workbench logic.
- **`Resources`**: Contains icons, translation files, and other resources.
- **`__init__.py`**: Indicates that the directory is a Python package.

### Creating Required Files

#### Init.py

This file will contain code to initialize your workbench when FreeCAD starts.

```python
def initialize():
    """ This function is executed when FreeCAD starts """
    import MyWorkbench
    MyWorkbench.initialize()

def FreeCADInit():
    """ This function is necessary for FreeCAD to recognize the Workbench """
    return "MyWorkbench"
```

#### InitGui.py

This file manages the GUI components of your Workbench.

```python
class MyWorkbenchWorkbench (Workbench):
    """ The Workbench class provides necessary information """
    def __init__(self):
        self.__class__.MenuText = "My Workbench"
        self.__class__.ToolTip = "My custom workbench for FreeCAD"
        self.__class__.Icon = ":/icons/myworkbench_icon.svg"
        
    def Initialize(self):
        """ This function is executed when the Workbench gets activated """
        import MyWorkbenchGui
        self.appendToolbar("My Tools", ["MyTool"])
        self.appendMenu("My Workbench", ["MyTool"])

    def Activated(self):
        """ Execute this function when the workbench is loaded """
        from PySide import QtCore, QtGui

    def Deactivated(self):
        """ Execute this when switching away from the workbench """
        pass

    def GetClassName(self):
        return "Gui::PythonWorkbench"

Gui.addWorkbench(MyWorkbenchWorkbench())
```

### Adding Tools

To add tools to your workbench, you'll need to create corresponding Python scripts and define them in the GUI initialization file.

#### Example Tool Script: MyTool.py

```python
class MyTool:
    """ Tool to perform a specific task """
    def Activated(self):
        print("MyTool Activated")
        
    def GetResources(self):
        return {
            'Pixmap' : 'path/to/your/tool/icon.svg',
            'MenuText': 'My Tool',
            'ToolTip' : 'Description of My Tool'
        }

Gui.addCommand('MyTool', MyTool())
```

#### Add the Tool to InitGui.py

Update your `InitGui.py` to include the new tool in a toolbar or menu.

```python
self.appendToolbar("My Tools", ["MyTool"])
self.appendMenu("My Workbench", ["MyTool"])
```

### Resource Files

Place your icons, translations, and other resources under the `Resources` directory. Use them in your tool and workbench definitions as shown in the scripts above.

### Packaging and Installing Your Workbench

- **Packaging**: Zip your `MyWorkbench` folder and its contents.
- **Installing**: Unzip the folder into FreeCAD's `Mod` directory.

When users start FreeCAD, your Workbench should now appear in the Workbench list.

### Advanced Features

- **Custom Dialogs**: Use `PySide` or `PyQt` to create custom dialogs.
- **Complex Algorithms**: Integrate third-party Python libraries to add advanced capabilities.

### Troubleshooting and Debugging

- **Log Files**: Check FreeCAD log files for any error messages.
- **Debugging**: Use Python's `pdb` module or IDE debugging tools.

### Conclusion

You've now created a basic custom Workbench for FreeCAD. Feel free to extend it with more features and share it with the community!
```

This guide should help developers create their own FreeCAD Workbench from scratch, providing all the necessary steps and code snippets.

**Comment by lukh** [comment]
Hello there,
I wrote my first workbench for FreeCAD (pure python) and found it quite hard to get access to "good doc" (ie, up to date, following standard python rules, etc)
The default template give informations, but not how to create a task for instance, 
The wiki is... (sorry :s )  complicated to get to, too many pages, in different sections, (maybe the scripting page are okay, but I couldn't find infos like; where to put the code snippet, I had to figure it out with some existing WB, (but not specifically up-to-date... or Draft Workbench, that uses a lot of shared / polymorphism code, so hard to follow when you want understand how a WB works...

I want to help, I can help @joelgraff  on the template, or on the wiki; I would probably need validation that the code I wrote is good enough for that

(BTW: learning FreeCAD for one year or so, love it, works like a charm :) ),

Bests

**Comment by luzpaz** [comment]
@lukh It would be very welcome update and change. Can you connect to the forum or discord chat and start a discussion relate to this?

**Comment by PhoneDroid** [comment]
Will be addressed via the Addon Academy.
